### DISCUSSION TITLE
Ideas for Future Features and Improvements

### DISCUSSION THREAD

**Discussion by ro-sjda42** [CombobulativeDesigns/GPTWebNovelTranslator#18] [Ideas]
I'm opening this post mostly to keep track of my own ideas for making the program better.

## Ideas

- [x] Translate metadata.
- [x] Parallelize API requests.
- [ ] Cost calculation.
- [ ] Cost report.
- [ ] Separate markdown and epub steps.
- [x] Novel wide terms sheet.
- [ ] DB option for storing Novel objects.
- [x] Specification of target language.
- [x] Importing and exporting term sheets.
- [x] Modularizing novel origin.
- [x] Supporting multiple sources.
- [ ] Local source.

### Translate metadata

Self explanatory. All metadata including table of contents, titles, and author is currently untranslated. I really should translate that as well.

### Parallelize API requests

Requests for the API currently run one after the other in series. There are requests that are independent of each other and can be run concurrently. The time we could save with this is gigantic.

### Cost calculation

I think I should be able to calculate approximate cost of a translation while making a few educated gueses. This will give the user a valuable resource in choosing what to translate.

### Cost report

Once a translation has been completed, we should have all info necesary to calculate the exact cost of the process. I should present that to the user.

### Separate markdown and epub steps

If I separate the writing of the markdown and epub files, I could implement the markdown as an editing step. Users could easily edit the translated sections to fix any errors the model might have done. Then they would be able to convert it to an epub file from the same program.

### Novel wide terms sheet.

Current terms sheet is implemented to be contextual to the sub chapter. Terms that appear early might fall out of use and get pruned out of future terms sheets as the novel progresses. This would leave the model with the task of re translating that term which could end in inconsistencies with earlier sub chapters.

If I make a novel wide term sheet, then select the terms to pass to the model based on their appearance in the current chunk of text, and then filtered to the top 15 according to their points, then we would in theory reduce inconsistencies.

### DB option for storing Novel objects

Don't really need everything to operate on a json object that could. MongoDB could be a nice alternative.

### Specification of target language

This tools works using the GPT API. As long as GPT is capable of translating to that language, the tool should be able to utilize that capability.

### Importing and exporting term sheets

Term sheets should be able to be shared and personalized. This requires the capability of exporting and importing them.

### Modularizing novel origin

Currently Syosetu ncode is hardcoded as the only origin for scraping. Modularizing the origins would pave the way for new sources.

### Supporting multiple sources

Once modularization of origins is complete, we can start developing more modules for other sources.

### Local source

Origin module for importing a local file into the tool for translating. Requires defining an imput format for the file. Thinking about a mix of markdown and yaml like epub uses.

**Comment by TidusFH** [comment]
Do you think it would be possible to make it translate .txt,.epubs,etc would make it easier to translate from non-jp sites like korean novels.

Is the term sheet purpouse to optimize translations? I've played with openai and it seems to be possible to optimize translations with it,i've been collection general terms of jp,kr and cn web novels to try to optimize when i translate them

**Reply by ro-sjda42** [reply]
Should be quite posible. Translation is done on the Novel objects, and with a target chapters/sub chapters specification. This basically means that, as long as we can convert some input to the Novel format the tool uses, then we can pass it to the translator.

Novel objects have: 
- novel_code: str
- title : str
- author : str
- description : str
- original_language : str
- chapters: list[Chapter]

Chapter objects have:
- chapter_index: int
- name: str
- sub_chapters: list[SubChapter]

And SubChapter objects have:
- sub_chapter_index: int
- name: str
- contents: str

They have more stuff, but that is handled by the tool so no reason to focus on it for the moment. As long as we can convert any input to this format, then we can translate it. Then we just need a way to insert it in the local storage of the app but that's trivial at this point.

Also, about the term sheet. It's not really about optimizing. The API has limits on token counts for each prompt, so we have to make multiple requests for long texts. But this can cause problems with continuity. One prompt could translate the name たくと (no idea of its kanji writing so forgive me) as Takuto, and another as Takt. Term's sheet provide context of previous translations relevant to the current text.

It can also provide better translations for terms than what the model can generate. This is specially useful if you want to make the translation cheaper, since you can translate using GPT-3.5 but with the term sheet generated by GPT-4, which improves quality a little bit.

We could make it so the app allows users to input their own term sheets. We can even allow exporting for sharing, modifying, and then importing.

Another idea is have a global term sheet which is combined with the novel term sheet before translation.

Ideas, ideas...

**Reply by TidusFH** [reply]
It might be easier to optimize when that chatgpt plugin comes out, but for now if its about making translations better the most interesting idea would be to find a way to make a basic general sheet,if more people get interested could have sheets that fit different genres more.

Do you happen to know if there is a API for Bing? I ask because it uses chatgpt-4 but the thing is that it  have acess to the internet so theorically we could make it acess a page that have terms for all kind of novels,it would make it much easier tbh.

**Reply by ro-sjda42** [reply]
Term sheets are built in an incremental way. I shove a chunk of the text to the API and ask it to create a list of relevant terms while focusing on proper nouns and technical terms. Then feed the response to the TermSheet object for the novel which parses the string and updates the sheet with the new terms found by the model. 

The tool goes through all targeted chapter splitting them into more chunks, and proceeds to do the same thing with all of them. Repeated terms are skipped, which means the earlier translation will win.

So if you import a term sheet with pre set terms, the tool will not only use your translations, but it will also not overwrite them in the future. And as the model translates, it will add more relevant terms to the sheet, making it grow more specific to the novel by the chunk.

All we need is a way to import a pre created sheet into the tool and we'll be good to go.

I've just added 4 more items to the "ideas" list in relation to importing/exporting term sheets, and multiple origins. 

Thank you for your input.

And about the API for Bing, maybe this [https://github.com/acheong08/EdgeGPT](https://github.com/acheong08/EdgeGPT) or this [https://github.com/stefanopochet/Chat-GPT-4-Bing-AI-API](https://github.com/stefanopochet/Chat-GPT-4-Bing-AI-API)? Don't have any experience with them, but that's what a quick google search gave me.

**Comment by TidusFH** [comment]
THose seem to work they seem to use Bing Cookies to acess the API the EdgeGPT seems pretty straightforward using an code to acess the API through bing chat cookies it seems to use a third party called rapidAPI , i have both the upgraded version on OpenAI and the Bing one i will mess around and see ihow big the difference in translations are from chatp gpt-3 because i honestly don't know if its worth to even bother.

Your sheet idea seems good, when i mentioned using bing AI to acess a sheet site was thinking more in terms of making use of the AI ability of reasoning,  JP is very hard even and could be messed up even with a good term sheet, being able to use of the AI reasoning  in a more efficient way would definetly save time and create a more accurate translation but for now just trying to implement the sheets should be good enough

**Reply by ro-sjda42** [reply]
I believe Bing uses GTP-4 right? Is it better than OpenAI's GPT-4 API? Because you can use that one with this tool. It just requires an OpenAI account with a GPT-4 API invitation. It's really not that hard to get, they gave me one in less than a week. The API is paid though, and GPT-4 is 15 times as expensive as GPT-3.5 with a cost of 0.03 usd per 1k tokens.

I have some good news though. I just modified the scraper for the tool. Made it modular so we can start adding more origins of novels. Right now I have both Syosetu NCode and Novel18 working on some branch of the repo. I plan on adding a local file origin, and another web one, before I release a new version of the tool.

Still need to figure out the input format for local origin. Was thinking of just using a reduced json format of the models contained in the tool. They aren't that hard to use if you just start from a template and copy paste the contents of the chapters in their respective places.

For the third web origin, I have no idea what to choose. Do you have any recomendations?

**Reply by TidusFH** [reply]
Do you have preference on the language of the novel origin? If its JP kakuyomu.jp is othet place you can get a bunch of free WN,some are also on syosetu but not all of them

The API being paid and that exppensive is a bummer, maybe there could be other options? I will check it out if i find something i will say something

**Reply by ro-sjda42** [reply]
It'd be nice for prices to lower as the time passes and the technology improves. 

I focused on OpenAI's GPT models because of how good they were at composing natural sounding text. I think they are a good way of getting over the artifacts found in other methods of machine translation. I do need to study Bing and see if they are a good alternative. Even if they are the same thing underneath.

And no preference on language of the novel origin. In fact, another language would be a good test of the models capabilities as a sort of universal translator. 

I'll test how kakuyomu.jp works with the novel format. Should be able to make it fit someway, and create an origin for that. Right now adding origins is simple, just need to figure out what elements to scrape.

**Reply by TidusFH** [reply]
Well for CN there is www.jjwxc.net its  a pretty good site it even have tags for search and stuff, for KR the only one i know its booktoki283.com its a very messy site though

I'm currently trying to translate a CN novel while structure and overall meaning are kept the AI still struggle with names its one of the things a term sheet could help with

**Reply by ro-sjda42** [reply]
Oh nice! That's some good info. I'll check those sites and see about adding origins for them.

Also, I just implemented an export/import function for term sheets (and novels themselves). User are now capable of sharing, or even improving sheets manually without much issue by exporting them, editing them, and importing them. Nice. 

And if you import a pre made sheet before your first translation, those terms will not be overwriten by the AI on term sheet generation. So now you can prepare sheets for things like genres and load them into the novel before translation for influencing the AI's translation. The thing you talked about a few comments before.

Update later today or tomorrow. Depends on how much it takes me to figure out what to scrape in those sites.
